rand_os/src/solarish.rs

```diff
@@ -110,18 +112,48 @@ impl OsRngImpl for OsRng {
     }
 }
 
-fn getrandom(buf: &mut [u8], blocking: bool) -> libc::c_long {
-    extern "C" {
-        fn syscall(number: libc::c_long, ...) -> libc::c_long;
+#[cfg(target_os = "illumos")]
+type GetRandomFn = unsafe extern fn(*mut u8, libc::size_t, libc::c_uint)
+    -> libc::ssize_t;
+#[cfg(target_os = "solaris")]
+type GetRandomFn = unsafe extern fn(*mut u8, libc::size_t, libc::c_uint)
+    -> libc::c_int;
+
+// Use dlsym to determine if getrandom(2) is present in libc.  On Solarish
+// systems, the syscall interface is not stable and can change between
+// updates.  Even worse, issuing unsupported syscalls will cause the system
+// to generate a SIGSYS signal (usually terminating the program).
+// Instead the stable APIs are exposed via libc.  Cache the result of the
+// lookup for future calls.  This is loosely modeled after the
+// libstd::sys::unix::weak macro which unfortunately is not exported.
+fn fetch() -> Option<GetRandomFn> {
+    static FPTR: AtomicUsize = AtomicUsize::new(1);
+
+    if FPTR.load(Ordering::SeqCst) == 1 {
+        let name = "getrandom\0";
+        let addr = unsafe {
+            libc::dlsym(libc::RTLD_DEFAULT, name.as_ptr() as *const _) as usize
+        };
+        FPTR.store(addr, Ordering::SeqCst);
     }
 
-    const SYS_GETRANDOM: libc::c_long = 143;
+    let ptr = FPTR.load(Ordering::SeqCst);
+    unsafe {
+        mem::transmute::<usize, Option<GetRandomFn>>(ptr)
```

Comment:
Since at this point ptr is either 0 (i.e. NULL) or the address of `getrandom(2)`, I used the nullable pointer optimization as described in the rustnomicon (the FFI chapter).

Reply:
Sure, that's what the optimisation does, but you *are* relying on the enum layout not changing (which the rustnomicon gives good reason for leaving unspecified) — though granted the layout for `Option<T>` with `T` being non-nullable is unlikely to change. All the same, there is zero reason to depend on this; as I demonstrated you can achieve exactly the same thing without dependence on the layout (or making use of `transmute`, which should be seen as a last resort). Don't worry about "optimisation" here; it will likely compile down to the same thing.

Reply:
I'm confused then.  The rustnomicon _explicitly_ calls this out as correct and the way to pass around a pointer to a C function (that's either defined or NULL) when dealing with FFI.  In fact it instructs one to pass an Option<T> _to a C function_ when the T is a function pointer or NULL -- that always is going to require the layout of Option<T> to be a certain way or the call will have undefined results.

Reply:
Good point, the type correspondence is correct thus the layout must also be fixed. I guess the point about data layouts of `enum` don't apply to the special case of `Option` with a non-nullable value then.

In which case the `transmute` is correct, though that still doesn't mean it's the preferred method of conversion (transmute has fewer error checks than the pointer cast method I suggested).